File: packages/sdk/src/retry.ts
// Retry / backoff policy for SDK requests.
//
// One chokepoint — every public client method goes through
// `withRetry(...)`. The policy:
//
//   * Up to N attempts (default 3, configurable per-client).
//   * Exponential backoff with ±25% jitter, capped at maxDelayMs.
//   * Retries 429 + 5xx (502/503/504 by default).
//   * Honors `Retry-After` header when present (seconds or HTTP-date).
//   * Idempotent methods (GET / DELETE) always retried; POSTs only
//     retried when the route is on the explicit allowlist below.
//   * Propagates AbortSignal — an in-flight backoff sleep aborts
//     immediately if the caller cancels.
//
// The retry/stream/cancel design lives in
// docs/development/sdks/NODE_SDK_IMPLEMENTATION.md §5.

import { TextralApiError, TextralRetryExhausted } from './errors.js';

export interface RetryInfo {
  /** 1-indexed. The first attempt is `1`; the first *retry* is `2`. */
  attempt: number;
  /** HTTP status when present (null for transport-level failures). */
  status: number | null;
  /** Computed sleep before the next attempt, in ms. */
  delayMs: number;
  /** The thrown value from the failed attempt. */
  error: unknown;
}

export interface RetryPolicy {
  /** Total attempts including the first. Default 3 = original + 2
   *  retries. Set to 1 to disable retries entirely. */
  maxAttempts: number;
  /** Base delay before the first retry. */
  initialDelayMs: number;
  /** Cap on the per-attempt sleep. */
  maxDelayMs: number;
  /** HTTP statuses that trigger a retry. Anything else falls
   *  straight through. */
  retryOn: number[];
  /** Optional telemetry hook. Fires on every retry-eligible failure
   *  *before* the backoff sleep. */
  onRetry?: (info: RetryInfo) => void;
  /** Allow specific POST routes to opt into retry. Format:
   *  "POST /v1/path". The default allowlist is the set of routes
   *  with server-side dedupe semantics — see DEFAULT_RETRY below. */
  retryOnPostMethods?: string[];
}

/** Routes whose POST handlers are idempotent server-side and
 *  therefore safe to retry on transient failure. Adding to this
 *  list requires the server-side guarantee — DON'T extend without
 *  checking the route handler. Each entry's reason:
 *
 *    /v1/auth/redeem
 *      Single-use token; consumed-token retry returns the same
 *      response. Safe.
 *
 *    /v1/auth/register, /v1/auth/recover
 *      Always return 202; safe to re-fire (worst case: two emails
 *      to the same user, which is already rate-limited server-side).
 *
 *    /v1/ingest/bulk
 *      Hashed by `client_request_id` for 24h. Safe when the caller
 *      passes one.
 *
 *    /v1/documents/{id}/uploads/{u}/finalize
 *      Server is idempotent on `(document_id, content_hash)`. Safe.
 *
 *    /v1/ingest/bulk/{id}/finalize
 *      Per-file finalize is idempotent. Safe to retry the whole call.
 */
// Patterns are matched against the route AFTER the client's
// `normalizePath()` has replaced ULID-shaped segments with `{id}`.
// Adding here requires server-side dedupe / no-op-on-duplicate
// semantics — see the comment block above.
const DEFAULT_IDEMPOTENT_POSTS = [
  'POST /v1/auth/redeem',
  'POST /v1/auth/register',
  'POST /v1/auth/recover',
  'POST /v1/ingest/bulk',
  'POST /v1/ingest/bulk/{id}/finalize',
  'POST /v1/documents/{id}/uploads/{id}/finalize',
] as const;

export const DEFAULT_RETRY: RetryPolicy = {
  maxAttempts: 3,
  initialDelayMs: 200,
  maxDelayMs: 5_000,
  retryOn: [429, 502, 503, 504],
  retryOnPostMethods: [...DEFAULT_IDEMPOTENT_POSTS],
};

/** No retries. Pass this to a client's `retry:` field to opt out. */
export const NO_RETRY: RetryPolicy = {
  maxAttempts: 1,
  initialDelayMs: 0,
  maxDelayMs: 0,
  retryOn: [],
};

export interface RequestContext {
  /** HTTP method, uppercase. */
  method: string;
  /** Route path used for the idempotent-POST allowlist match. The
   *  caller normalizes path params (e.g. `/v1/documents/{id}/uploads/{u}/finalize`).
   *  Match is exact-string against `retryOnPostMethods`. */
  routePattern: string;
}

/** Run `fn` with retry semantics. Returns whatever `fn` returns;
 *  throws one of three things:
 *
 *    1. `fn`'s underlying error verbatim — when the error is NOT
 *       retryable (wrong status code, non-idempotent POST, abort).
 *    2. `TextralRetryExhausted` — when the error WAS retryable but
 *       we've used up the configured `maxAttempts`. Wraps the last
 *       error in `.cause`.
 *    3. `AbortError` — when the caller's signal aborts at any
 *       point (between attempts or during a backoff sleep).
 *
 *  These three are deliberately distinct so call sites can branch
 *  cleanly on intent. */
export async function withRetry<T>(
  fn: () => Promise<T>,
  policy: RetryPolicy,
  ctx: RequestContext,
  signal?: AbortSignal,
): Promise<T> {
  let lastError: unknown;
  for (let attempt = 1; attempt <= policy.maxAttempts; attempt++) {
    if (signal?.aborted) {
      throw signal.reason ?? new DOMException('Aborted', 'AbortError');
    }
    try {
      return await fn();
    } catch (err) {
      lastError = err;
      // Non-retryable → throw original error.
      if (!isRetryableError(err, policy, ctx)) throw err;
      // Retryable but out of attempts → throw exhausted wrapper.
      if (attempt >= policy.maxAttempts) {
        throw new TextralRetryExhausted(policy.maxAttempts, err);
      }
      // Otherwise: sleep with backoff, then loop.
      const status = extractStatus(err);
      const delayMs = computeDelay(attempt, policy, err);
      policy.onRetry?.({ attempt, status, delayMs, error: err });
      await sleep(delayMs, signal);
    }
  }
  // Unreachable — the loop body either returns or throws on every
  // iteration. Defensive throw keeps TS happy + catches a future
  // refactor that drops the throw.
  throw new TextralRetryExhausted(policy.maxAttempts, lastError);
}

function isRetryableError(
  err: unknown,
  policy: RetryPolicy,
  ctx: RequestContext,
): boolean {
  // Aborts are never retried — caller asked us to stop.
  if (err instanceof Error && err.name === 'AbortError') return false;

  // POST methods need to be on the idempotent allowlist.
  if (ctx.method === 'POST') {
    const allowlist = policy.retryOnPostMethods ?? [];
    const target = `${ctx.method} ${ctx.routePattern}`;
    if (!allowlist.includes(target)) return false;
  }

  if (err instanceof TextralApiError) {
    return policy.retryOn.includes(err.status);
  }
  // Transport-level failure (e.g., DNS, connection-reset) →
  // retryable by default.
  return true;
}

function extractStatus(err: unknown): number | null {
  if (err instanceof TextralApiError) return err.status;
  return null;
}

function computeDelay(
  attempt: number,
  policy: RetryPolicy,
  err: unknown,
): number {
  // Honor Retry-After when present (the server is telling us when
  // to come back; ignore our own backoff math in that case).
  if (err instanceof TextralApiError && err.retryAfter !== undefined) {
    return Math.min(err.retryAfter * 1000, policy.maxDelayMs);
  }
  // Exponential backoff with ±25% jitter.
  const base = Math.min(
    policy.initialDelayMs * Math.pow(2, attempt - 1),
    policy.maxDelayMs,
  );
  const jitter = base * 0.25 * (Math.random() * 2 - 1);
  return Math.max(0, Math.round(base + jitter));
}

/** Cancellable sleep. Resolves after `ms`; rejects with AbortError
 *  if the signal aborts before then. */
function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(signal.reason ?? new DOMException('Aborted', 'AbortError'));
      return;
    }
    const timer = setTimeout(() => {
      signal?.removeEventListener('abort', onAbort);
      resolve();
    }, ms);
    const onAbort = () => {
      clearTimeout(timer);
      reject(signal?.reason ?? new DOMException('Aborted', 'AbortError'));
    };
    signal?.addEventListener('abort', onAbort, { once: true });
  });
}

/** Parse a `Retry-After` header (seconds or HTTP-date) into a
 *  delay in seconds. Exported so the client's response handler
 *  can stash it on the TextralApiError it builds. Returns
 *  `undefined` when the header is missing or unparseable. */
export function parseRetryAfter(value: string | null | undefined): number | undefined {
  if (!value) return undefined;
  const trimmed = value.trim();
  // Plain seconds.
  const asSeconds = Number(trimmed);
  if (Number.isFinite(asSeconds) && asSeconds >= 0) return asSeconds;
  // HTTP-date.
  const ts = Date.parse(trimmed);
  if (!Number.isNaN(ts)) {
    const seconds = Math.max(0, Math.round((ts - Date.now()) / 1000));
    return seconds;
  }
  return undefined;
}

File: packages/sdk-python/src/textral/_retry.py
"""Retry policy for SDK requests. 1:1 mirror of
@textral/sdk's retry.ts so the two SDKs behave identically under
transient failure:

  * Up to N attempts (default 3).
  * Exponential backoff with ±25% jitter, capped at max_delay_ms.
  * Honors Retry-After when present.
  * GET/DELETE always retried; POST only on the idempotent
    allowlist (server-side dedupe makes those safe).
  * Raises `TextralRetryExhausted` when the policy gives up.
  * Returns the original exception unchanged for non-retryable
    errors.

Cancellation: `signal` parameter is not used (Python uses
`asyncio.wait_for` / native `KeyboardInterrupt` at the call-site
level — see _http.py for the integration with httpx timeouts).
"""

from __future__ import annotations

import asyncio
import random
import re
import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from email.utils import parsedate_to_datetime
from typing import TypeVar

from .errors import TextralAPIError, TextralRetryExhausted

T = TypeVar("T")


# Routes whose POST handlers are server-side idempotent and
# therefore safe to retry. Patterns are matched against the route
# AFTER `_normalize_path()` has replaced ULID-shaped segments with
# `{id}`. Adding here requires server-side dedupe / no-op-on-
# duplicate semantics.
DEFAULT_IDEMPOTENT_POSTS: tuple[str, ...] = (
    "POST /v1/auth/redeem",
    "POST /v1/auth/register",
    "POST /v1/auth/recover",
    "POST /v1/ingest/bulk",
    "POST /v1/ingest/bulk/{id}/finalize",
    "POST /v1/documents/{id}/uploads/{id}/finalize",
)


@dataclass
class RetryInfo:
    """Passed to the on_retry hook on every retry-eligible failure
    BEFORE the backoff sleep."""

    attempt: int
    """1-indexed; first attempt is 1, first retry is 2."""
    status: int | None
    """HTTP status (None for transport-level errors)."""
    delay_ms: int
    """Sleep before the next attempt."""
    error: BaseException


@dataclass
class RetryPolicy:
    """Configurable retry behavior. Mirrors the TS-side defaults."""

    max_attempts: int = 3
    initial_delay_ms: int = 200
    max_delay_ms: int = 5_000
    retry_on: tuple[int, ...] = (429, 502, 503, 504)
    on_retry: Callable[[RetryInfo], None] | None = None
    retry_on_post_methods: tuple[str, ...] = field(
        default_factory=lambda: DEFAULT_IDEMPOTENT_POSTS
    )


DEFAULT_RETRY = RetryPolicy()
"""Idiomatic default — 3 attempts, exponential backoff, retry on
429/5xx, idempotent-POST allowlist as documented in the design."""


# Sentinel for "no retry at all". Pass `retry=NO_RETRY` (or
# `retry=None`) to the Client constructor to disable retries.
NO_RETRY = RetryPolicy(
    max_attempts=1,
    initial_delay_ms=0,
    max_delay_ms=0,
    retry_on=(),
    retry_on_post_methods=(),
)


@dataclass(frozen=True)
class RequestContext:
    """The route context for one HTTP call. The retry layer reads
    `method` + `route_pattern` to decide whether the call is
    eligible for retry."""

    method: str
    """Uppercased HTTP method."""
    route_pattern: str
    """The path with ULID segments normalized to `{id}`. Used for
    idempotent-POST allowlist matching."""


_ULID_SEGMENT = re.compile(r"/[a-z]{2,5}_[A-Z0-9]{26}")


def normalize_path(path: str) -> str:
    """Replace prefixed-ULID segments with `{id}` and drop query
    string. Used to build a stable route pattern for the retry
    allowlist regardless of which document/version/job id the path
    references."""
    base = path.split("?", 1)[0]
    return _ULID_SEGMENT.sub("/{id}", base)


def parse_retry_after(value: str | None) -> float | None:
    """Parse a `Retry-After` header. Accepts seconds-as-int or
    HTTP-date. Returns seconds (float), or None when missing or
    unparseable."""
    if not value:
        return None
    s = value.strip()
    # Plain seconds.
    try:
        seconds = float(s)
        return max(0.0, seconds)
    except ValueError:
        pass
    # HTTP-date.
    try:
        when = parsedate_to_datetime(s)
        delta = when.timestamp() - time.time()
        return max(0.0, delta)
    except (TypeError, ValueError):
        return None


def _is_retryable_error(
    err: BaseException, policy: RetryPolicy, ctx: RequestContext
) -> bool:
    # Cancellation / explicit interrupt is never retried.
    if isinstance(err, (asyncio.CancelledError, KeyboardInterrupt)):
        return False

    # POST methods need to be on the idempotent allowlist.
    if ctx.method == "POST":
        target = f"{ctx.method} {ctx.route_pattern}"
        if target not in policy.retry_on_post_methods:
            return False

    # API-shaped error: retry on declared statuses.
    if isinstance(err, TextralAPIError):
        return err.status_code in policy.retry_on

    # Transport-level (httpx.ConnectError, TimeoutException, etc.)
    # → retryable by default.
    return True


def _compute_delay(attempt: int, policy: RetryPolicy, err: BaseException) -> int:
    """Either Retry-After (when the server told us when to come
    back) or exponential backoff with ±25% jitter."""
    if isinstance(err, TextralAPIError) and err.retry_after is not None:
        return min(int(err.retry_after * 1000), policy.max_delay_ms)
    base = min(
        policy.initial_delay_ms * (2 ** (attempt - 1)),
        policy.max_delay_ms,
    )
    jitter = base * 0.25 * (random.random() * 2 - 1)
    return max(0, int(base + jitter))


# ── Sync + async runners. Sister functions; the policy semantics
# are identical, only the awaiter differs.


def with_retry(
    fn: Callable[[], T],
    policy: RetryPolicy,
    ctx: RequestContext,
) -> T:
    """Synchronous retry wrapper.

    Returns whatever `fn()` returns; raises the underlying error
    when not retryable; raises `TextralRetryExhausted` when the
    policy gives up after max_attempts attempts on a retryable
    error."""
    last_error: BaseException | None = None
    for attempt in range(1, policy.max_attempts + 1):
        try:
            return fn()
        except BaseException as err:
            last_error = err
            if not _is_retryable_error(err, policy, ctx):
                raise
            if attempt >= policy.max_attempts:
                raise TextralRetryExhausted(policy.max_attempts, err) from err
            status = err.status_code if isinstance(err, TextralAPIError) else None
            delay_ms = _compute_delay(attempt, policy, err)
            if policy.on_retry:
                policy.on_retry(
                    RetryInfo(
                        attempt=attempt,
                        status=status,
                        delay_ms=delay_ms,
                        error=err,
                    )
                )
            time.sleep(delay_ms / 1000.0)

    # Defensive — the loop body always returns or raises.
    if last_error is not None:
        raise TextralRetryExhausted(policy.max_attempts, last_error) from last_error
    raise RuntimeError("with_retry exited without a result")  # pragma: no cover


async def with_retry_async(
    fn: Callable[[], Awaitable[T]],
    policy: RetryPolicy,
    ctx: RequestContext,
) -> T:
    """Async retry wrapper — same semantics, awaitable runner."""
    last_error: BaseException | None = None
    for attempt in range(1, policy.max_attempts + 1):
        try:
            return await fn()
        except BaseException as err:
            last_error = err
            if not _is_retryable_error(err, policy, ctx):
                raise
            if attempt >= policy.max_attempts:
                raise TextralRetryExhausted(policy.max_attempts, err) from err
            status = err.status_code if isinstance(err, TextralAPIError) else None
            delay_ms = _compute_delay(attempt, policy, err)
            if policy.on_retry:
                policy.on_retry(
                    RetryInfo(
                        attempt=attempt,
                        status=status,
                        delay_ms=delay_ms,
                        error=err,
                    )
                )
            await asyncio.sleep(delay_ms / 1000.0)

    if last_error is not None:
        raise TextralRetryExhausted(policy.max_attempts, last_error) from last_error
    raise RuntimeError("with_retry_async exited without a result")  # pragma: no cover


